### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by gonziesc**
travis
## Summary
- Auth implementation
- Middleware validated token date
- User can invalidate all endpoints, logout and renew
- Token expires after 30 days

## Trello Card
https://trello.com/c/CO1J8y6M/5-autenticacion

### CODE REVIEW

File: app/middlewares.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        req.user = u;
+        req.accessToken = accessToken;
+        next();
+      } else {
+        res.status(401);
+        res.end();
+      }
+    });
+  } else {
+    res.status(401);
```
Comment (mishuagopian) [inline app/middlewares.js]: We can do something like:
```js
next(error.unauthorized);
```

File: app/controllers/users.js [hunk 2]
```diff
@@ -8,13 +10,72 @@ exports.create = (req, res, next) => {
...
+  res.status(200);
+  res.end();
+};
+
+exports.renew = (req, res) => {
+  const accessToken = req.accessToken;
+  const user = req.user;
+  if (req.user.email === accessToken.email) {
+    if (moment().isBefore(accessToken.expirationDateWarning)) {
+      res.status(403);
```
Comment (mishuagopian) [inline app/controllers/users.js]: Remember user `next(errors.warningDateNotReached)`

File: app/middlewares.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    userManager.getByMail(accessToken.email).then((u) => {
+      if (u && moment().isBefore(accessToken.expirationDate)) {
+        if (moment().isAfter(accessToken.expirationDateWarning)) {
+          res.set(sessionManager.WARNING_HEADER_NAME, true);
+        }
+        req.user = u;
+        req.accessToken = accessToken;
+        next();
+      } else {
+        res.status(401);
```
Comment (mishuagopian) [inline app/middlewares.js]: Use `next(error.someError)`

File: app/controllers/users.js [hunk 4]
```diff
@@ -8,13 +10,72 @@ exports.create = (req, res, next) => {
...
+      res.send('Warning expiration date has not been reached');
+    } else {
+      res.status(200);
```
Comment (mishuagopian) [inline app/controllers/users.js]: You should renew tokens expiration date here.

Comment (mishuagopian) [inline app/controllers/users.js]: Remember also checking for maximum useful date.

File: app/services/sessionManager.js [hunk 5]
```diff
@@ -1,16 +1,42 @@
...
-exports.generateAccessToken = (email) => {
-  const expiringDate = new Date();
-  expiringDate.setDate(expiringDate.getDate() + 7);
-  const claims = { mail: email,
-    exp: expiringDate
+const MAXIMUM_USEFUL_DAYS = 30;
+const EXPIRATION_DATE = 2;
+const EXPIRATION_DATE_WARNING = 10;
+
+const maximumUsefulDate = () => {
```
Comment (mishuagopian) [inline app/services/sessionManager.js]:
```js
const maximumUsefulDate = () => moment().add(MAXIMUM_USEFUL_DAYS, 'days');
```

File: app/services/sessionManager.js [hunk 6]
```diff
@@ -1,16 +1,42 @@
...
+  return moment().add(MAXIMUM_USEFUL_DAYS, 'days');
+};
+const expirationDate = () => {
+  return moment().add(EXPIRATION_DATE, 'days');
```
Comment (mishuagopian) [inline app/services/sessionManager.js]: same

File: app/services/sessionManager.js [hunk 7]
```diff
@@ -1,16 +1,42 @@
...
+};
+const expirationDateWarning = () => {
+  return moment().add(EXPIRATION_DATE_WARNING, 'seconds');
```
Comment (mishuagopian) [inline app/services/sessionManager.js]: same

File: app/services/sessionManager.js [hunk 8]
```diff
@@ -1,16 +1,42 @@
...
+};
+exports.generateAccessTokenWithRenewId = (user, renewId) => {
+  return exports.encode({
+    email: user.email,
+    maximumUsefulDate: maximumUsefulDate(),
+    expirationDate: expirationDate(),
+    expirationDateWarning: expirationDateWarning(),
+    id: user.id,
```
Comment (mishuagopian) [inline app/services/sessionManager.js]: This is unused here

File: app/controllers/users.js [hunk 9]
```diff
@@ -8,13 +10,72 @@ exports.create = (req, res, next) => {
...
+        }
+      } else {
+        next(errors.invalidUser);
+      }
+    });
+  }
+};
+
+exports.invalidateAll = (req, res) => {
+  const user = req.user;
```
Comment (mishuagopian) [inline app/controllers/users.js]: You should change some value here so all tokens that have that value are invalid now. If not, you are doing nothing here.

File: app/controllers/users.js [hunk 10]
```diff
@@ -1,14 +1,17 @@
...
   errors = require('../errors');
 
 exports.create = (req, res, next) => {
   const user = req.body ? {
     firstName: req.body.firstName,
     lastName: req.body.lastName,
     password: req.body.password,
-    email: req.body.email
+    email: req.body.email,
+    verificationCode: 1
```
Comment (mishuagopian) [inline app/controllers/users.js]: This is because we can't create a hash, right?

File: app/controllers/users.js [hunk 11]
```diff
@@ -18,3 +21,59 @@ exports.create = (req, res, next) => {
...
+  const email = req.body.email;
+  const errorsSend = userService.validateEmail(email);
+  if (errorsSend.length) {
+    return next(errors.clientError(errorsSend));
+  } else {
+    userService.getByMail(email).then((u) => {
+      if (u && u.password === password) {
+        const accessToken = session.generateAccessToken(u);
+        res.status(200);
+        res.set(session.HEADER_NAME, accessToken);
```
Comment (mishuagopian) [inline app/controllers/users.js]: You have no need of setting the header. Just return the access token object.

File: app/controllers/users.js [hunk 12]
```diff
@@ -18,3 +21,59 @@ exports.create = (req, res, next) => {
...
+};
+
+exports.renew = (req, res, next) => {
+  const accessToken = req.accessToken;
+  const user = req.user;
+  if (req.user.email === accessToken.email) {
+    if (moment().isBefore(accessToken.expirationDateWarning)) {
+      return next(errors.warningDateNotReached);
+    } else {
+      if (moment().isBefore(accessToken.maximumUsefulDate)) {
```
Comment (mishuagopian) [inline app/controllers/users.js]: This is not OK. `maximumUsefulDate` should be used as a date from which a user can't renew the given token.

File: app/controllers/users.js [hunk 13]
```diff
@@ -18,3 +21,59 @@ exports.create = (req, res, next) => {
...
+  user.verificationCode = 0;
+  user.save();
+  res.status(200);
+  res.end();
+};
+
+exports.renew = (req, res, next) => {
+  const accessToken = req.accessToken;
+  const user = req.user;
+  if (req.user.email === accessToken.email) {
```
Comment (mishuagopian) [inline app/controllers/users.js]: You should check for the `renew` param. The user must send it to renew its access.

File: app/middlewares.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  sessionManager = require('./services/sessionManager'),
+  errors = require('./errors'),
+  orm = require('./orm').models;
+
+exports.secure = (req, res, next) => {
+  const auth = req.headers[sessionManager.HEADER_NAME];
+  if (auth) {
+    const accessToken = sessionManager.decode(auth);
+    userManager.getByMail(accessToken.email).then((u) => {
+      if (u && u.verificationCode === accessToken.verificationCode && moment().isBefore(accessToken.expirationDate)) {
```
Comment (mishuagopian) [inline app/middlewares.js]: I'd also check we are before `maximumUsefulDate`

File: app/controllers/users.js (hunk 10 above)
Comment (gonziesc) [inline app/controllers/users.js]: Yes

File: app/controllers/users.js [hunk 15]
```diff
@@ -18,3 +21,57 @@ exports.create = (req, res, next) => {
...
+  const accessToken = req.accessToken;
+  const user = req.user;
+  if (req.renew_id === accessToken.renewId || moment().isBefore(accessToken.maximumUsefulDate)) {
+    const renew = req.renewId;
+    if (req.user.email === accessToken.email) {
+      if (moment().isBefore(accessToken.expirationDateWarning)) {
+        return next(errors.warningDateNotReached);
+      } else {
+        accessToken.expirationDate = session.expirationDate();
+        res.status(200);
```
Comment (mishuagopian) [inline app/controllers/users.js]: You must return the modified `accessToken` here (obviously encoded)

File: app/middlewares/auth.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+const moment = require('moment'),
+  userManager = require('../services/users'),
+  sessionManager = require('../services/sessionManager'),
+  orm = require('../orm').models,
+  errors = require('../errors');
+
+exports.secure = (req, res, next) => {
```
Comment (mishuagopian) [inline app/middlewares/auth.js]: Remove this file. It's unused.

**Pull request merged**